### PR TITLE
logging: add simplelog sample

### PR DESCRIPTION
**Pull request by broady**

### CODE REVIEW

**Comment by broady** [comment]
cc @jmdobry @jonparrott @pfritzsche

**Comment by broady** [comment]
Tests coming now. Take a look and see if this is reasonable first, though?

File: logging/simplelog/simplelog.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		// For example, if Flush() failed.
+		log.Printf("client.OnError: %v", err)
+	}
+	// [END setup]
+
+	log.Print("Writing some log entries.")
+	writeEntry(client)
+	structuredWrite(client)
+
+	// It takes some time for the log entries to appear.
```
Comment (jonparrott) [inline logging/simplelog/simplelog.go]: In Python we did these as separate functions/commands, which avoids the necessity of a sleep in the code.

Comment (broady) [inline logging/simplelog/simplelog.go]: I saw. I like it. This seems OK, too. You just "go run" and it does everything.

Comment (jonparrott) [inline logging/simplelog/simplelog.go]: Sure - either way. I'm neurotic about sleeps in code, but it's fine if you're cool with. :)

File: logging/simplelog/simplelog.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
+// iample simplelog writes some entries, lists them, then deletes the log.
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: Sample simplelog...

File: logging/simplelog/simplelog.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	"cloud.google.com/go/preview/logging"
+
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	// [END imports]
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Println("Usage: simplelog <project-id>")
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: We use an env variable for the project name in some references, an argument in some other. Why not using GCLOUD_PROJECT env var everywhere?

File: logging/simplelog/simplelog.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	}
+}
+
+func writeEntry(client *logging.Client) {
+	// [START write_log_entry]
+	const name = "log-example"
+	logger := client.Logger(name)
+	defer logger.Flush() // Ensure the entry is written.
+
+	stdlog := logger.StandardLogger(logging.Info)
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: s/stdlog/infolog would be nicer. We are expecting users to construct multiple loggers for each log level.

File: logging/simplelog/simplelog.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+			entry.Payload)
+	}
+
+	log.Print("Deleting log.")
+	if err := deleteLog(client); err != nil {
+		log.Fatalf("Could not delete log: %v", err)
+	}
+}
+
+func writeEntry(client *logging.Client) {
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: Should we call this writeWithStdLogger something?

File: logging/simplelog/simplelog.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	// [END write_log_entry]
+}
+
+func deleteLog(client *logging.Client) error {
+	ctx := context.Background()
+
+	// [START delete_log]
+	const name = "log-example"
+	if err := client.DeleteLog(ctx, name); err != nil {
+		return err
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: Wanna log.Fatal here and not returning error from deleteLog?

File: logging/simplelog/simplelog.go (hunk 1 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: Done

File: logging/simplelog/simplelog.go (hunk 6 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: It's for a snippet, so I thought `return err` might be better without any context. What do you think?

File: logging/simplelog/simplelog.go (hunk 4 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: Done. Changed the message a little, too.

File: logging/simplelog/simplelog.go (hunk 2 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: Done

File: logging/simplelog/simplelog.go (hunk 6 above)
Comment (rakyll) [inline logging/simplelog/simplelog.go]: yea, ok, we don't know what user want to do with the error.

File: logging/simplelog/simplelog.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
+// Sample simplelog writes some entries, lists them, then deletes the log.
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: Missing license header.

File: logging/simplelog/simplelog.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	// [END imports]
+)
+
+func main() {
+	if len(os.Args) != 3 {
+		if len(os.Args) == 2 {
+			usage("Missing command.")
+		} else {
```
Comment (rakyll) [inline logging/simplelog/simplelog.go]: You don't need the else. usage exists.

if len(os.Args) == 2 {
   usage("...")
}

usage("")

Comment (rakyll) [inline logging/simplelog/simplelog.go]: Actually, let's make this easier.

if len(os.Args) == 2 {
   usage("...")
}

if len(os.Args) != 3 {
    usage("")
}

**Comment by rakyll** [comment]
LGTM with a nits.

File: logging/simplelog/simplelog.go (hunk 7 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: Done

File: logging/simplelog/simplelog.go (hunk 8 above)
Comment (broady) [inline logging/simplelog/simplelog.go]: Yes, the second one is better. It isn't obvious `usage` exits, so I added the else block.

**Comment by broady** [comment]
Added tests. Have a look?

They'll be quite flaky, unfortunately. Let's see how bad it is in practice. Might be OK for now (like pubsub).

File: logging/simplelog/simplelog_test.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+	time.Sleep(5 * time.Second)
+
+	entries, err := getEntries(client, tc.ProjectID)
+	if err != nil {
+		t.Fatalf("getEntries: %v", err)
+	}
+
+	if got, want := len(entries), 2; got != want {
+		t.Fatalf("len(entries): got %d; want %d", got, want)
```
Comment (rakyll) [inline logging/simplelog/simplelog_test.go]: len(entries) = %d; want %d

No need for got when you don't have to explain what you got in English.

File: logging/simplelog/simplelog_test.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	wantContain := map[string]*logging.Entry{
+		"Anything":                            entries[0],
+		"The payload can be any type!":        entries[0],
+		"infolog is a standard Go log.Logger": entries[1],
+	}
+
+	for want, entry := range wantContain {
+		msg := fmt.Sprintf("%s", entry.Payload)
+		if !strings.Contains(msg, want) {
+			t.Errorf("want %q to contain %q", msg, want)
```
Comment (rakyll) [inline logging/simplelog/simplelog_test.go]: Is entry.Payload encoded in JSON? Why can't we assert the JSON body rather than a strings.Contains?

**Comment by rakyll** [comment]
LGTM with a few nits.

File: logging/simplelog/simplelog_test.go (hunk 10 above)
Comment (broady) [inline logging/simplelog/simplelog_test.go]: Nope, it comes back as a proto struct (also, undocumented).

**Pull request merged**